### PR TITLE
Adding complex expressions projections for Subquery

### LINKED ISSUE #9576: Keep track of common sub-expression across logical plan nodes
### Is your feature request related to a problem or challenge?

_No response_

### Describe the solution you'd like

Currently, common `CommonSubexprEliminate` `LogicalPlan` optimizer rule analyzes common sub-expressions in a query. Then caches, common sub-expression by adding a `LogicalPlan::Projection` if it thinks this is beneficial.
As an example, following query
```sql
SELECT c3+c4, SUM(c3+c4) OVER(order by c3+c4)
FROM t
```
generates following `LogicalPlan`:
```
Projection: t.c3 + t.c4, SUM(t.c3 + t.c4) ORDER BY [t.c3 + t.c4 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
--WindowAggr: windowExpr=[[SUM(CAST(t.c3 + t.c4t.c4t.c3 AS t.c3 + t.c4 AS Int64)) ORDER BY [t.c3 + t.c4t.c4t.c3 AS t.c3 + t.c4 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW AS SUM(t.c3 + t.c4) ORDER BY [t.c3 + t.c4 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW]]
----Projection: t.c3 + t.c4 AS t.c3 + t.c4t.c4t.c3, t.c3, t.c4
------TableScan: t projection=[c3, c4]
```
where `t.c3+t.c4` is calculated once in the `Projection` then referred by subsequent `WindowAggr` as a column.

However, following query:
```sql
SELECT c3+c4, SUM(c3+c4) OVER()
FROM t
```
generates following `LogicalPlan`:
```
Projection: t.c3 + t.c4, SUM(t.c3 + t.c4) ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
--WindowAggr: windowExpr=[[SUM(CAST(t.c3 + t.c4 AS Int64)) ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING]]
----TableScan: t projection=[c3, c4]
```
instead we could generate following plan:
```
Projection: col(t.c3 + t.c4), SUM(t.c3 + t.c4) ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
--WindowAggr: windowExpr=[[SUM(CAST(col(t.c3 + t.c4) AS t.c3 + t.c4 AS Int64)) ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING]]
----Projection: t.c3 + t.c4 AS col(t.c3 + t.c4)
------TableScan: t projection=[c3, c4]
```
If were to keep track of common sub expression counts globally across different nodes in the `LogicalPlan`. This will enable us to generate better `LogicalPlan`s.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #9576 

## What changes are included in this PR?
Added a top-down-top tree-node traverse in all plan, find complex_expressions in all LogicalPlan, try to add an extra Projection if this expression is needed in upper layer.

### CODE REVIEW

**Review by waynexia — COMMENTED** [review]
Thanks for taking this @Lordworms 

One situation I came up with that makes eliminating common expr across plans tricky is how to keep track of column references. Consider a plan like:

```
Projection: a+1
--Projection: a+1 as a
----TableScan
```

Where two `a+1` are different things actually, since two `a` are not the same thing. The result is wrong if we replace them with a pre-computed common expr.

**Comment by Lordworms** [comment]
> Thanks for taking this @Lordworms
> 
> One situation I came up with that makes eliminating common expr across plans tricky is how to keep track of column references. Consider a plan like:
> 
> ```
> Projection: a+1
> --Projection: a+1 as a
> ----TableScan
> ```
> 
> Where two `a+1` are different things actually, since two `a` are not the same thing. The result is wrong if we replace them with a pre-computed common expr.

Thanks, I'll record this and fix in the following PR

File: datafusion/sqllogictest/test_files/project_complex_sub_query.slt
Comment (Lordworms) [reply datafusion/sqllogictest/test_files/project_complex_sub_query.slt:32]: Ok

File: datafusion/optimizer/src/optimize_projections.rs
Comment (Lordworms) [reply datafusion/optimizer/src/optimize_projections.rs:736]: Got it

File: datafusion/sqllogictest/test_files/project_complex_sub_query.slt
Comment (mustafasrepo) [inline datafusion/sqllogictest/test_files/project_complex_sub_query.slt:42]: I double checked this plan. It seems that expression `t.c3+t.c4` in the top projection still is a binary expression (Unfortunately from display it is not visible whether it is column or binary expression). We should have convert it to its pre-evaluated column version: `Col(t.c3+t.c4)`. If that was the case, I would expect to see following logical plan
```
Projection: t.c3 + t.c4, SUM(t.c3 + t.c4) ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
--WindowAggr: windowExpr=[[SUM(CAST(t.c3 + t.c4 AS Int64)) ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING]]
----Projection: t.c3 + t.c4 AS t.c3 + t.c4
------TableScan: t projection=[c3, c4]
```
where columns `t.c3` and `t.c4` are pruned at the first projection (from bottom). Since they are no longer needed by subsequent stages. I will try to solve this problem. If I can come up with a solution I will let you know.

File: datafusion/sqllogictest/test_files/project_complex_sub_query.slt
Comment (mustafasrepo) [inline datafusion/sqllogictest/test_files/project_complex_sub_query.slt:32]: since `demo.csv` file is not added to this PR. Following tests fails to produce `PhysicalPlan`. I think, it is better to add this file also. We can still test `LogicalPlan`s in the test by using the flag `set datafusion.explain.logical_plan_only = true;`

Comment (Lordworms) [reply datafusion/sqllogictest/test_files/project_complex_sub_query.slt:32]: OK, I'll add that

File: datafusion/sqllogictest/test_files/project_complex_sub_query.slt
Comment (mustafasrepo) [reply datafusion/sqllogictest/test_files/project_complex_sub_query.slt:42]: Feel free to go ahead.

Comment (Lordworms) [reply datafusion/sqllogictest/test_files/project_complex_sub_query.slt:42]: Sure

**Comment by Lordworms** [comment]
I think right now it is fine, I'll run the tests
image

**Comment by Lordworms** [comment]
@mustafasrepo Hi If you have a moment, I would be honored to have your review on my updated one.

**Comment by mustafasrepo** [comment]
> @mustafasrepo Hi If you have a moment, I would be honored to have your review on my updated one.

Sure, I will look into this PR.

File: datafusion/sqllogictest/test_files/project_complex_sub_query.slt
Comment (mustafasrepo) [reply datafusion/sqllogictest/test_files/project_complex_sub_query.slt:42]: I think same problem still occurs in the `LogicalPlan::Window` where inner expression of the `Cast` is binary expression. I would expect `Projection` immediately top of the `TableScan` to only include expression: `t.c3 + t.c4 AS t.c3 + t.c4` because. Subsequent stages doesn't refer `t.c3` or `t.c4`. If you rewrite the expression for window also (as in projection). We should get that plan. 
In short we have to rewrite expressions that contribute to common expression count during `f_down` pass.

**Comment by Lordworms** [comment]
Thanks @mustafasrepo , I double-checked the results and it meet the requirements. I actually have one question, why we could still get the proper answer with eliminating the upper ProjectionExec? I think it is ready to be merged now.

**Comment by mustafasrepo** [comment]
> Thanks @mustafasrepo , I double-checked the results and it meet the requirements. I actually have one question, why we could still get the proper answer with eliminating the upper ProjectionExec?(I think it is optimized since another logical plan optimization would eliminate the same Projection) I think it is ready to be merged now. Also, I think we could use similar way to track Aggregate, I could make an extra ticket here and solve it in the coming PR.

Exactly, the `optimize_projections` rule (for `LogicalPlan`) removes the unnecessary `LogicalPlan::Projection`s from the plan if those projections have same input and same output schema. In this example, this is the case also. `LogicalPlan::window` has already desired schema for the user input.

**Review by mustafasrepo — APPROVED** [review]
LGTM!. Thanks @Lordworms for this work. It is really nice to collaborate with you. Another look is appreciated though. If any reviewer has time.

**Comment by Lordworms** [comment]
> LGTM!. Thanks @Lordworms for this work. It is really nice to collaborate with you. Another look is appreciated though. If any reviewer has time.

Thank you!

File: datafusion/optimizer/src/common_subexpr_eliminate.rs
Comment (andygrove) [inline datafusion/optimizer/src/common_subexpr_eliminate.rs:420]: This comment seems to be incomplete?

File: datafusion/optimizer/src/common_subexpr_eliminate.rs
Comment (andygrove) [inline datafusion/optimizer/src/common_subexpr_eliminate.rs:869]: Is there a tracking issue for this that we could refer to in the comment?

Comment (Lordworms) [reply datafusion/optimizer/src/common_subexpr_eliminate.rs:869]: I'll add that

File: datafusion/optimizer/src/common_subexpr_eliminate.rs
Comment (Lordworms) [reply datafusion/optimizer/src/common_subexpr_eliminate.rs:420]: Sorry for that, I'll complete it

**Comment by alamb** [comment]
I am trying to clean up our old PRs -- it looks to me like  @mustafasrepo  approved and asked for another reviewer https://github.com/apache/datafusion/pull/9719#pullrequestreview-1967920447. However it isn't clear to me if this PR needs another review before merge or if that is a nice to have. 

 I don't think we can merge this PR without some work to resolve the conflicts, but I don't want to waste people's time resolving conflicts if it isn't ready to merge.

Please let us know your thoughts

**Comment by mustafasrepo** [comment]
> However it isn't clear to me if this PR needs another review before merge or if that is a nice to have.
While asking for another review, What I had in mind was "nice to have". I think, once conflicts are resolved we can merge this PR as is.

**Review by waynexia — APPROVED** [review]
Thank you all for working on this ❤️ 

I've reviewed it again, it's mostly LGTM. +1 for merging this once the conflict and remaining comments are solved.

**Comment by alamb** [comment]
Marking this one as a draft as I think it needs to have conflicts resolved before we can merge it

**Pull request closed without merging**
